### DISCUSSION TITLE
Thanks ratatui, plus rendering best practices

### DISCUSSION THREAD

**Discussion by agl-alexglopez** [ratatui/ratatui#579] [Discussion]
I just finished up converting a maze visualization program I have been working on for a while to a TUI application ([check it out if you'd like](https://github.com/agl-alexglopez/maze-tui)).
 
![demo](https://github.com/ratatui-org/ratatui/assets/87786254/6e3842a3-2a66-4663-9d65-a39eb82a713b)

Previously, I used escape codes, cleared the terminal screen, and moved the cursor or printed directly to the command line as needed to achieve what I wanted. Ratatui is way better than that so thank you for working on it! I did have a question about the rendering model and best practices given my focus for this project is smooth animations.

Because this project is about mazes, early on I chose to encode instructions for what should be printed to the screen in the bits of a maze square. For animations, while the algorithms do their work they will ask for the cursor at a position on the screen equal to the maze row/column coordinate and the bits encode what is printed, like certain colors or wall shapes for the mazes, and then they sleep for a set time. I had to do this because printing the entire maze to the command line for every change in the maze would be terrible, slow, and have flickering. The lines and animation looks somewhat choppy in the GIF but on an actual screen different animation speeds can look quite smooth.

When changing over to ratatui I kept cursor movement as the primary mechanism for animation. This relates to discussion #409 about partial rendering, but I chose to render the minimum possible widgets at any time so I wouldn't have to redraw the screen every frame. I don't render any widgets during the animations, instead using crossterm to move the cursor and execute print commands. My preconceived notion about the terminal is that it is slow, so #409 surprised me and now I am curious if I should consider rendering in a different way. I read over the design documentation about the Buffer, Cell, and View model. How much control can I have over these lower level elements? I am not sure what the best approach to animation is with ratatui. The priority would definitely be smooth animations with no flickering and control over the speed of the action. Let me know if you have any thoughts. Also, this is my first Rust program and I don't have any TUI/GUI experience so please feel free to correct me about anything I mentioned that seems off. Thanks again!

**Comment by joshka** [comment]
First up. This is pretty damn cool!

> My preconceived notion about the terminal is that it is slow, so https://github.com/ratatui-org/ratatui/discussions/409 surprised me and now I am curious if I should consider rendering in a different way.

I suspect that I was wrong with some of my terminal speed assumptions. I tried animating the colors_rgb example and saw it maxing out around 3fps. That had a custom background and foreground color for each cell though, so 10k (200x50 cols/rows) changes per frame each with rgb ansi commands. I'm guessing that this would be close to the upper bound of the complexity of most applications.

So there's definitely some work to do there to make it super fast. There's some techniques that we could implement from https://www.textualize.io/blog/7-things-ive-learned-building-a-modern-tui-framework/ that might help - particularly single write and sychronized output protocol.


>When changing over to ratatui I kept cursor movement as the primary mechanism for animation. This relates to discussion https://github.com/ratatui-org/ratatui/discussions/409 about partial rendering, but I chose to render the minimum possible widgets at any time so I wouldn't have to redraw the screen every frame. I don't render any widgets during the animations, instead using crossterm to move the cursor and execute print commands.

Did you try out the approach of rendering a full widget? On each frame, the backend will compare the current buffer to the new buffer and only write the commands necessary to update the screen, so in effect we're probably already doing the same thing you're doing here. I suspect the best approach would be having a widget that has a maze grid as state that's kept between renders and updated by the maze algorithm and keeping the widget as a simple write to the buffer. Assuming the code to convert your maze to a buffer is fairly fast (it should be quick to call `buf.get_mut().set_symbol("x").set_fg(...).set_bg(...)` in a tight loop), then I don't think there would be much difference in your manual crossterm code vs just using ratatui for this. (Of course happy to be shown wrong on this - theory and practice are often not 100% aligned).

> I read over the design documentation about the Buffer, Cell, and View model. How much control can I have over these lower level elements? I am not sure what the best approach to animation is with ratatui. The priority would definitely be smooth animations with no flickering and control over the speed of the action.

Widget::render allows you to pretty much do what you'd like with the buffer for every cell, and we're actually going to expose this directly on `Frame` in the upcoming release if you just want to hit that directly from the terminal.draw call.
https://github.com/ratatui-org/ratatui/blob/ad2dc5646dae04fa5502e677182cdeb0c3630cce/src/terminal.rs#L646

> For animations, while the algorithms do their work they will ask for the cursor at a position on the screen equal to the maze row/column coordinate and the bits encode what is printed, like certain colors or wall shapes for the mazes, and then they sleep for a set time. I had to do this because printing the entire maze to the command line for every change in the maze would be terrible, slow, and have flickering. 

One approach to doing animation that you might consider is to treat time as the input to some function that generates the state, and then rendering at ~60fps (or whatever you can achieve) rather than treating the next frame as the input and sleeping for some time. I.e. instead of asking "when do I need to render this next piece of information" to make it look good, ask "given that it's 16ms later, what should I render now".

This allows you to do things like draw things even more smoothly than you currently do. E.g. the examples that splat some maze corridor to the maze could instead splat a small portion of that chunk instead of the whole chunk, the search patterns could use unicode characters like 1/8 blocks to enter a cell rather than a full block. This also would make the keyboard more responsive than places where you're rendering a full row of characters before coming back to the main loop. It probably would also allow you to more nuanced speed increases as 1 frame no longer has to equal exactly 1 step of the algorithm.

I'm a big proponent of measuring when talking performance. Something that would be worth doing would be profiling your code with [cargo-flamegraph](https://github.com/killercup/cargo-flamegraph). This will help show which parts of the application are the actual hotspots. E.g.:

```shell
cargo flamegraph --root --release --bin run_tui
```
![flamegraph](https://github.com/ratatui-org/ratatui/assets/381361/4ee89b6d-a7d4-47db-ad53-c7cffa5f8975)

Note that github removes the interactivity of the svg. If you load the file locally, you get an interactive svg to drill into rather than a static one - e.g.:

<img width="497" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/5fed618c-19d6-4bf8-b40f-24fce33538b6">
And zooming in on the Terminal::draw:
<img width="1727" alt="Screenshot 2023-10-20 at 3 06 06 PM" src="https://github.com/ratatui-org/ratatui/assets/381361/e7e769a3-22b9-4bb6-b351-b5c2b5ab6748">

If you wanted to compare the two approaches easily what about making the app take command line parameters like `run_maze` does, and just rendering a full maze and quitting when done.

If you do run into perf issues that you can definitely correlate to being ratatui introduced rather than the underlying problem, this seems like a good project to help figure them out.

**Reply by agl-alexglopez** [reply]
Ok, this is great! I want to make sure to thank you for the response before trying some things out. I'll start a new branch to explore animating the maze with the buffer approach. 

I really like the idea of getting more nuanced Unicode characters involved with an fps rendering approach, if possible. Early on I experimented with variable box widths to signify thread solver traffic and it was _ok_. I think full width variable height could look good though. This would improve algorithms like breadth first search a lot, where a batch update of the state per X milliseconds would make the spread look much more natural and less choppy. Lots to think about, thanks!

**Reply by kdheepak** [reply]
The content in this link is relevant https://ratatui.rs/concepts/rendering-under-the-hood.html (and I think some of the discussion from here should be ported to that link too).

Also, to add to @joshka's excellent answer, I did some tests related to this a few years ago in Julia. I found that if you are flushing after every write (i.e. unbuffered writes), it is faster to build a sequence of ascii characters as a string and write it out in one go if the amount of content being written out is large e.g. if every pixel in the terminal (i.e. 100% of the pixels) has changed. If only around 30% of the pixels have changed, I found it was a wash. If fewer than 10% of the pixels had changed, it was faster to move the cursor to the exact pixel and make changes there.

However, when using crossterm, ratatui calls `queue!(stdout, ...)` which doesn't actually write anything to the terminal. At the end of the `draw()` call, ratatui calls `terminal.flush()` which then writes everything out. I haven't done exact performance tests with this but I believe this uses buffered writes to the terminal which should make it more performant. 

Also, for performance, you can add a fps widget that should tell you till you are dropping frames. I've found that Rust / LLVM are quite good at optimizing a lot of things. There are some widgets in Ratatui that require allocations which is not ideal (but with a Widget reuse PR that might change) but if you don't have allocations (you can write a custom widget or write directly to the buffer with the latest alpha release of Ratatui) you should get pretty good performance.

**Reply by joshka** [reply]
> However, when using crossterm, ratatui calls queue!(stdout, ...) which doesn't actually write anything to the terminal. At the end of the draw() call, ratatui calls terminal.flush() which then writes everything out. I haven't done exact performance tests with this but I believe this uses buffered writes to the terminal which should make it more performant.

I must admit I haven't actually looked too deeply at that code. But that makes a lot of sense.

**Reply by joshka** [reply]
>  really like the idea of getting more nuanced Unicode characters involved with an fps rendering approach, if possible

Something else that you can do is use half block characters to effectively give yourself 2 square characters per rectangular cell to use for maze building rather than one. We use that to good effect in `demo2` and the `colors_rgb` example, and it's added to the Canvas as an option in the alpha release.

**Reply by agl-alexglopez** [reply]
>  it is faster to build a sequence of ascii characters as a string and write it out in one go if the amount of content being written out is large e.g. if every pixel in the terminal (i.e. 100% of the pixels) has changed. If only around 30% of the pixels have changed, I found it was a wash.

Thanks for those figures! Maze building algorithms could change very little at a time (like in a recursive backtracker) or many cells (like in Kruskal's algorithm). Maze solving has a hard cap of 50% of Cells to work with total because paths make up half the maze and the amount of Cells changing at any given time would be far less than the total amount. I think overall the percentages would be low but I will have to consider.

> Something else that you can do is use half block characters to effectively give yourself 2 square characters per rectangular cell to use for maze building rather than one.

Good one! I can picture this 2 for one strategy exactly. And when you solve the maze you just change the background color to fill the gaps and that represents the solver path, instead of changing the entire square. I'll try to get something working. It's lucky timing also with this new release. I think some of those changes will be helpful.

**Reply by kdheepak** [reply]
After typing this comment I was curious about this, so I did a little profiling. Here's the [game of life ratatui tui](https://github.com/kdheepak/game-of-life) samply profiler output (from https://github.com/mstange/samply):

<img width="1204" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/21756173-6cb8-4a10-9749-5d3073eed460">

98% of samples appear to be in the `write` in the `queue` function. The crossterm documentation has these lines:

```
    /// * When `flush` is called manually on the given type implementing `io::Write`.
    /// * The terminal will `flush` automatically if the buffer is full.
    /// * Each line is flushed in case of `stdout`, because it is line buffered.
```

https://github.com/crossterm-rs/crossterm/blob/08762b3ef4519e7f834453bf91e3fe36f4c63fe7/src/command.rs#L74-L76

I have to read through the crossterm source to see what exactly happens for `stdout` vs `stderr` but it does seem like for the exact same terminal window size, using `stdout` instead of `stderr` made it faster. Visually I could see a difference, as well as the samply profiler output shows fewer samples in the `write` in the `queue` function of crossterm:

<img width="1206" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/79ee5dbb-44e7-43e9-94a6-77ce8b208d80">

This particular application of game of life has a lot of diffs in each frame and very little time is spent in Ratatui and a lot of time is spent actually writing to the terminal.

For example, if I decrease the size of the terminal (1/4th the size as previous runs), and don't change anything in the code, here's what I get as the profiler result

<img width="1206" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/0021a4da-e2fb-4745-81fc-fe7bbe016ba7">

**Reply by kdheepak** [reply]
To confirm, digging through `write_command_ansi` appears to end up in system calls:

stderr `Print` command in crossterm:

<img width="1205" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/77128cc1-b7f9-4670-a905-9f3a7b5ea762">

stdout `Print` command in crossterm:

<img width="1206" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/253238fa-26bb-4b45-bf34-51dc7d6acecd">

**Reply by kdheepak** [reply]
If every pixel is different, Ratatui's crossterm backend at the very least still does as good a job as it can (imo):

https://github.com/ratatui-org/ratatui/blob/ad2dc5646dae04fa5502e677182cdeb0c3630cce/src/backend/crossterm.rs#L134-L137

i.e. it only moves the cursor if the next diff position is not immediate the next pixel. 

My takeaway is that if you are changing nearly every pixel or large number of randomly positioned pixels in the TUI every frame, then Ratatui's diffing algorithm is going to do slightly more work. Still, the biggest bottleneck is actually writing to the terminal which is a system call, and there's only so much you can do in that situation when dealing with terminals. Unless something like [Synchronized output](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036) is implemented in crossterm, there's not much Ratatui would be able to do about it I imagine. 

Ratatui's diffing algorithm is efficient and you can probably rely on it instead of implementing your own with crossterm manually. Mind your allocations with Widgets and you should be able to squeeze out as much performance as you can for your application.

**Reply by agl-alexglopez** [reply]
@joshka I'm still working on comparing rendering techniques/performance. In the meantime I had to try that half block idea you mentioned for maze walls and thought you'd get a kick out of it. The geometry is really nice. It always bugged me how the cells are about twice as tall as they are wide making everything stretch, but this fixes that! Adding half blocks to the alpha release was a good choice.

https://github.com/ratatui-org/ratatui/assets/87786254/d23de967-d3fe-43c3-89b6-272d8c77457d

**Comment by joshka** [comment]
> My takeaway is that if you are changing nearly every pixel or large number of randomly positioned pixels in the TUI every frame, then Ratatui's diffing algorithm is going to do slightly more work. Still, the biggest bottleneck is actually writing to the terminal which is a system call, and there's only so much you can do in that situation when dealing with terminals. Unless something like [Synchronized output](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036) is implemented in crossterm, there's not much Ratatui would be able to do about it I imagine.

https://docs.rs/crossterm/0.27.0/crossterm/terminal/struct.BeginSynchronizedUpdate.html (was added in 0.26.1), but a quick test suggests it doesn't have any speed effect on the following:
<details><summary>example full screen animation</summary>

```rust
use std::{
    io::{stdout, Write},
    time::{Duration, Instant},
};

use anyhow::Result;
use crossterm::{
    event::{self, Event, KeyCode, KeyEventKind},
    terminal::{disable_raw_mode, enable_raw_mode, EnterAlternateScreen, LeaveAlternateScreen},
    ExecutableCommand,
};
use ratatui::{prelude::*, widgets::*};

fn main() -> Result<()> {
    App::run(stdout())
}

struct App {
    should_quit: bool,
    tick: usize,
    fps: f64,
    last_fps_tick: usize,
    last_fps_instant: Instant,
}

impl App {
    fn new() -> Result<Self> {
        Ok(Self {
            should_quit: false,
            tick: 0,
            fps: 0.0,
            last_fps_tick: 0,
            last_fps_instant: Instant::now(),
        })
    }

    pub fn run<W: Write>(writer: W) -> Result<()> {
        install_panic_hook();
        init_terminal()?;
        let mut terminal = Terminal::new(CrosstermBackend::new(writer))?;
        let mut app = Self::new()?;
        terminal.clear()?;
        while !app.should_quit {
            app.calculate_fps();
            terminal.draw(|frame| {
                app.render(frame);
            })?;
            app.handle_events()?;
            app.tick += 1;
        }
        restore_terminal()?;
        Ok(())
    }

    fn render(&mut self, frame: &mut Frame) {
        let layout = Layout::default()
            .direction(Direction::Vertical)
            .constraints([Constraint::Length(1), Constraint::Min(1)])
            .split(frame.size());
        let header = Paragraph::new(Text::raw(format!("FPS: {:.2}", self.fps)));
        frame.render_widget(header, layout[0]);
        let buf = frame.buffer_mut();
        let area = layout[1];
        for y in area.top()..area.bottom() {
            for x in area.left()..area.right() {
                buf.get_mut(x, y)
                    .set_symbol(match (x + y + self.tick as u16) % 3 {
                        0 => " ",
                        1 => ".",
                        2 => "o",
                        _ => unreachable!(),
                    });
            }
        }
    }

    fn handle_events(&mut self) -> Result<()> {
        if event::poll(Duration::from_millis(100))? {
            if let Event::Key(key) = event::read()? {
                if key.kind == KeyEventKind::Press && key.code == KeyCode::Char('q') {
                    self.should_quit = true;
                };
            }
        }
        Ok(())
    }

    fn calculate_fps(&mut self) {
        let elapsed_ticks = self.tick - self.last_fps_tick;
        let elapsed_duration = self.last_fps_instant.elapsed();
        if elapsed_ticks > 10 && elapsed_duration > Duration::from_secs(1) {
            self.fps = elapsed_ticks as f64 / elapsed_duration.as_secs_f64();
            self.last_fps_tick = self.tick;
            self.last_fps_instant = Instant::now();
        }
    }
}

fn init_terminal() -> Result<()> {
    enable_raw_mode()?;
    stdout().execute(EnterAlternateScreen)?;
    Ok(())
}

fn restore_terminal() -> Result<()> {
    stdout().execute(LeaveAlternateScreen)?;
    disable_raw_mode()?;
    Ok(())
}

fn install_panic_hook() {
    better_panic::install();
    let prev_hook = std::panic::take_hook();
    std::panic::set_hook(Box::new(move |info| {
        let _ = restore_terminal();
        prev_hook(info);
    }));
}
```
</details>

<img width="216" alt="image" src="https://github.com/ratatui-org/ratatui/assets/381361/3aa55f8d-078f-49c1-bd0f-bac1e97af175"> (iterm2, 2023 Macbook Pro M2 Max). ~~That's pretty :(~~

~~I suspect the fact that we're calling `queue!` for every cell rather than batching several cells into a single write might be the slow point - I don't have a profiler setup for rust yet, so that's just a guess.~~

Edit: I just spent way too long trying to work out why I was only seeing this hit 10fps... turns out it was PEBKAC - after bumping this down to a single ms, the fps jumps to 300+.
```rust
        if event::poll(Duration::from_millis(100))? {
```

https://github.com/ratatui-org/ratatui/pull/583 has a animated version of colors_rgb that runs at ~24-40 fps on my machine (I'm not sure where the variability in that comes from). That's updating the foreground and background color in every cell (except the top row). stderr runs at about 11-12 FPS.


https://github.com/ratatui-org/ratatui/assets/381361/5cd99151-0511-4501-a4b6-3704b82317dd

**Comment by agl-alexglopez** [comment]
I just finished completely refactoring my project to utilize ratatui frame based rendering for all animations, rather than crossterm cursor calls. These changes brought a number of benefits:

- Any animation can be paused, overlayed with popups, and resumed with no visual information lost.
- I can play the animations forward or in reverse. By focusing on deltas, or what squares I want to change given a certain amount of time has passed, I have fine grained control over how many Cells in the Buffer should change.
- The animation can be sped up or slowed down in real time as you are interacting with it.
- I'm no longer worried about performance because the overhead of my animations is much lower.

To clarify what I mean by the last point, by changing my approach to the program overall the rendering became very simple. Before, I had the maze building and solving algorithms call the cursor at coordinates on the screen and then sleep. The solvers in this program were multithreaded so this was a ton of system calls for the cursor and for thread management. When I switched over to frame based rendering I found the overhead of multiple threads trying to generate deltas to render along with the main application thread trying to manage everything hard to implement and I had little fine-grained control. 

Instead, I played out the entire maze building and solving algorithms with no display or I/O operations, instead recording how the maze changed in a data structure I wrote for that purpose. This allowed me to record the state of the maze before and after some action or a burst of actions was taken. I could then replay the algorithms with this playback information and map maze squares directly to Buffer Cells as @joshka suggested. Then, the entire render process is simple; render the entire maze on every frame and only provide deltas when a specified amount of time has passed. I then leave it to ratatui to manage updating, flushing, and switching out the Buffers behind the scenes.

I think I am only scratching the surface for what is possible with this approach and the number of Buffer Cells changing for any given update is quite small so performance is not a concern at all right now. I can try to generate a performance comparison between cursor and frame based rendering later but I can no longer think of a good reason to use the cursor directly when you have ratatui available to do rendering for you. I kept my old code around just in case I do further comparisons in the future, but if I were to use this new approach exclusively I could cut out thousands of lines of code complexity and repetition for cursor management. Even if you can find good reasons to manually manage the cursor I have found that much more is possible when you consider when and how many updates to provide for the ratatui render loop. I get that this is probably no surprise to you all but it was good for me to learn it first hand by switching a project over to your library. Thanks!

**Reply by joshka** [reply]
That's awesome news! Thanks for the write up on your progress. Sounds like you've got a bunch of new features to show off too :)

**Reply by kdheepak** [reply]
Sounds like it would make a good blog post! ;)
